feat(mcxa276): initial HAL import

bogdan-petru opened the pull request:
This pull request removes several configuration and documentation files related to project workflows, code ownership, and contribution guidelines. The changes primarily clean up CI/CD workflows and project metadata, likely in preparation for a significant refactor or migration.

### CI/CD Workflow Removals

* Deleted GitHub Actions workflows for code formatting, linting, documentation, feature checks, dependency vetting, and no-std compatibility, removing files such as `.github/workflows/check.yml`, `.github/workflows/cargo-vet.yml`, `.github/workflows/cargo-vet-pr-comment.yml`, and `.github/workflows/nostd.yml`. [[1]](diffhunk://#diff-55f05888fd854d9962bdd4d8ccf23a07b610990f58174397b2586b58f8937428L1-L169) [[2]](diffhunk://#diff-d24ae2a4b160e60c7a5ed188839fa8532414911e462088eac4d4f321716d521eL1-L53) [[3]](diffhunk://#diff-052caea641627229cd694d0a37f82dd0af5123d612dbc0147b73537f1c7b4692L1-L137) [[4]](diffhunk://#diff-d7c20ca608ea2811ed0d2eddeabd110d724cb949f40a3f8cce2877e8b077c59cL1-L30)

### Project Metadata and Contributor Files

* Removed the `CODEOWNERS` file, which previously listed repository maintainers.
* Deleted the `CONTRIBUTING.md` file, which contained contribution guidelines and licensing information.
* Removed the `CODE_OF_CONDUCT.md` file, which outlined expected community behavior and enforcement policies.

### Development Environment Configuration

* Deleted the `.vscode/settings.json` file, which provided editor and Rust analyzer settings for development.

### Build Configuration Addition

* Added a new `.cargo/config.toml` file with custom build targets, runner configuration, and cargo command aliases for embedded development workflows.

felipebalbi reviewed (changes requested):
It would be better if the PAC were its own repository. @jeffglaum , could you create a new repository for the PAC or give me permission to do so?

bogdan-petru committed:
feat(mcxa276): initial HAL import

jamesmunns commented:
Hey @bogdan-petru, I agree with @felipebalbi regarding the PAC, having it in a separate repo will help keep the HAL repo reasonably sized.

For the HAL work: Let me know how you'd like to collaborate and give any feedback. I'd say there's probably two ways that could make sense:

* We can merge in your current work as-is (after we split out the PAC), then I can open issues or a PR with notes, working on the main branch
* Or, we can iterate in PRs, either all at once (like the current work is), or one component at a time.

I'd probably prefer the first option, as I'd like to roll in a clocking API similar to https://github.com/tweedegolf/embassy-imxrt/pull/1, and it's likely we can do that in parallel to your ongoing driver work.

Let me know what you think!

bogdan-petru commented:
> Hey @bogdan-petru, I agree with @felipebalbi regarding the PAC, having it in a separate repo will help keep the HAL repo reasonably sized.
> 
> For the HAL work: Let me know how you'd like to collaborate and give any feedback. I'd say there's probably two ways that could make sense:
> 
> * We can merge in your current work as-is (after we split out the PAC), then I can open issues or a PR with notes, working on the main branch
> * Or, we can iterate in PRs, either all at once (like the current work is), or one component at a time.
> 
> I'd probably prefer the first option, as I'd like to roll in a clocking API similar to [tweedegolf/embassy-imxrt#1](https://github.com/tweedegolf/embassy-imxrt/pull/1), and it's likely we can do that in parallel to your ongoing driver work.
> 
> Let me know what you think!

Hi @jamesmunns , first option seems fine by me.

felipebalbi commented:
> > Hey @bogdan-petru, I agree with @felipebalbi regarding the PAC, having it in a separate repo will help keep the HAL repo reasonably sized.
> > For the HAL work: Let me know how you'd like to collaborate and give any feedback. I'd say there's probably two ways that could make sense:
> > 
> > * We can merge in your current work as-is (after we split out the PAC), then I can open issues or a PR with notes, working on the main branch
> > * Or, we can iterate in PRs, either all at once (like the current work is), or one component at a time.
> > 
> > I'd probably prefer the first option, as I'd like to roll in a clocking API similar to [tweedegolf/embassy-imxrt#1](https://github.com/tweedegolf/embassy-imxrt/pull/1), and it's likely we can do that in parallel to your ongoing driver work.
> > Let me know what you think!
> 
> Hi @jamesmunns , first option seems fine by me.

Here's the repo for the PAC:

https://github.com/OpenDevicePartnership/mcxa-pac

felipebalbi commented on README.md:
```diff
@@ -0,0 +1,340 @@
+
+**Problem:** The OSTIMER examples crashed during interrupt handling with a hardfault (SP=0x00000000). Investigation revealed the OS_EVENT interrupt vector was NULL in the vector table, causing the CPU to jump to address 0 when OSTIMER interrupts fired.
+
+**Root Cause:** The `mcxa276-pac/src/lib.rs` file (generated from the SVD file) was missing the `WAKETIMER0` interrupt handler declaration. This caused the `__INTERRUPTS` array to have an off-by-one error, placing OS_EVENT at IRQ 58 instead of the correct IRQ 57 position.
```
The "proper" way to fix this is to patch the SVD file. There are several options, but if you're using `svd2rust`, then `svdtools` is the easiest. Have a look at how we have been patching RT685 SVD.

- [PAC Repository](https://github.com/OpenDevicePartnership/mimxrt685s-pac)
- [Main patch file](https://github.com/OpenDevicePartnership/mimxrt685s-pac/blob/main/patch/MIMXRT685S_cm33.yaml)
- [Example patch](https://github.com/OpenDevicePartnership/mimxrt685s-pac/blob/main/patch/usart.yaml)

felipebalbi commented on SW-Content-Register.txt:
```diff
@@ -0,0 +1,76 @@
+                                 - Embassy integration (executors, time)
+                               Location: ./src, ./examples, ./build.rs, ./memory.x, ./ram.ld, ./defmt.x, ./.cargo/config.toml, ./run.sh, ./tools/
+                               Origin: OEMWCSE (MIT)
+                               URL: https://bitbucket.sw.nxp.com/scm/oemwcse/mcxa_rust.git
```
Maybe you'd like to avoid exposing internal URLs?

felipebalbi commented on SW-Content-Register.txt:
```diff
@@ -0,0 +1,76 @@
+                               Outgoing License: MIT OR Apache-2.0
+                               License File: (see crate metadata)
+                               Format: Rust source code (auto-generated PAC)
+                               Description: Auto-generated register mappings for MCXA276 peripherals (svd2rust). Includes device.x and interrupt vectors. This tree contains a small manual correction for a missing WAKETIMER0 vector/enum entry.
```
When sending the PAC PR to the other repo, could you include the SVD as well? Thank you

bogdan-petru committed:
feat(mcxa276): use external mcxa-pac; remove in-tree PAC; ADC: stop writing CMDL1.CTYPE (read-only)

bogdan-petru committed:
chore(mcxa276): pin mcxa-pac to a9dd33; docs: OS_EVENT/WAKETIMER0 fix now historical with SVD 25.06.00

bogdan-petru committed:
Fix uart_interrupt example: enable RX interrupts in peripheral

The uart_interrupt example was not echoing received characters because
the RX interrupt enable bit (RIE) was not being set in the LPUART
control register. The NVIC interrupt was configured and the handler
was installed, but the peripheral itself was not generating interrupts.

Added uart.enable_rx_interrupts() call to enable RX data interrupts
in the LPUART2 peripheral before entering the main loop.

jamesmunns commented on .cargo/config.toml:
```diff
@@ -0,0 +1,19 @@
+[target.thumbv8m.main-none-eabihf]
+runner = "./run.sh"
+# Use our custom ram.ld for RAM-execution
+rustflags = ["-C", "link-arg=-Tdefmt.x", "-C", "link-arg=-Tram.ld"]
```
At some point we should use a normal flash based memory.x

jamesmunns replied:
Opened as #3

jamesmunns commented on .cargo/config.toml:
```diff
@@ -0,0 +1,19 @@
+# Blocking run (uses run.sh). Good for one-terminal flash+run.
+run-defmt = ["run", "--features", "defmt defmt-rtt lpuart2", "--example", "hello"]
+# Non-blocking flash to RAM via J-Link; frees probe for RTT attach
+flash-nb = ["!./tools/run_jlink_noblock.sh", "target/thumbv8m.main-none-eabihf/debug/examples/hello", "1366:0101:000600110607", "MCXA276", "500"]
```
You may not want to hard-code the vid:pid:serial here

jamesmunns commented on .vscode/settings.json:
We probably want to restore some of these template files, as they are expected in ODP repos. CC @felipebalbi for exactly which ones are necessary.

felipebalbi replied:
I'm not super worried about vscode settings, though I don't understand why they were deleted. However, the entire `supply-chain` directory and github workflows should be left alone. Additionally, README should be in markdown format, as it was before.

jamesmunns replied:
Yeah, I just clicked on the first deleted file, there were a bunch of removals (probably just a hard switch over/copying in the files).

jamesmunns commented on Cargo.toml:
```diff
@@ -1,19 +1,121 @@
-# dependencies for all targets
+cortex-m = { version = "0.7", features = ["critical-section-single-core"] }
+cortex-m-rt = { version = "0.7", features = ["device"] }
+mcxa276-pac = { git = "https://github.com/bogdan-petru/mcxa-pac", features = ["rt"], rev = "a9dd330149912c26f252f0f3c419f6435c1216e2", version = "0.1.0" }
```
We may want to land the mcxa-pac PR first and use that as the git repo dep, or we can do a quick release to crates-io.

jamesmunns commented on Cargo.toml:
```diff
@@ -1,19 +1,121 @@
+mcxa276-pac = { git = "https://github.com/bogdan-petru/mcxa-pac", features = ["rt"], rev = "a9dd330149912c26f252f0f3c419f6435c1216e2", version = "0.1.0" }
+embedded-io = "0.6"
+# Optional defmt support
+defmt = { version = "0.3", optional = true }
```
```suggestion
defmt = { version = "1.0", optional = true }
```

felipebalbi replied:
Yeah, this really should be 1.0

jamesmunns commented on Cargo.toml:
```diff
@@ -1,19 +1,121 @@
+embedded-io = "0.6"
+# Optional defmt support
+defmt = { version = "0.3", optional = true }
+defmt-rtt = { version = "0.4", optional = true }
```
```suggestion
defmt-rtt = { version = "1.1", optional = true }
```

jamesmunns commented on LICENSE:
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2025 Open Device Partnership
+Copyright (c) 2025 OEMWCSE
```
Just wondering, what is "OEMWCSE"?

jamesmunns commented on README.md:
Maybe rename this README.md? Not a big deal either way :)

jamesmunns commented on src/adc.rs:
```diff
@@ -0,0 +1,375 @@
+            1 => {
+                adc.cmdl1().write(|w| {
+                    w.adch().variant(config.channel_number)
+                                          .mode().variant(config.conversion_resolution_mode)
```
might want to do one big `cargo fmt` commit before merging.

jamesmunns reviewed (commented):
Just a couple comments, nothing urgent to block the merge.

jamesmunns commented on run.sh:
At least on MacOS, I needed to make the following changes to make this work (just for the record, we don't necessarily have to fix this all now):

1. I needed to `chmod +x` this file to have `cargo run` work correctly
2. I needed to add `--connect-under-reset` to the probe-rs gdb command to avoid an error
3. `gdb-multiarch` doesn't seem to be easily installable with `brew` or similar, though I did replace it with `arm-none-eabi-gdb` and was able to get things working

Commit:
Reduce number of features

We don't need features for drivers that always exist.

Signed-off-by: Felipe Balbi <febalbi@microsoft.com>

Commit:
cargo +nightly fmt

Signed-off-by: Felipe Balbi <febalbi@microsoft.com>

felipebalbi commented:
pushed a few updates to the PR, things a little more manageable :) 

Does anybody oppose merging this as is and improving things in subsequent PRs?

jamesmunns commented:
No objections from me, did you want to get CI wired back up or restore any files in this PR? Or do in a follow-up PR?

felipebalbi commented:
> No objections from me, did you want to get CI wired back up or restore any files in this PR? Or do in a follow-up PR?

That's a good point. I'll do that before merging as well

Commit:
Reintroduce necessary files

Signed-off-by: Felipe Balbi <febalbi@microsoft.com>

Commit:
Vet dependencies

Signed-off-by: Felipe Balbi <febalbi@microsoft.com>

felipebalbi commented:
oh, right, github ci can't deal easily with private repositories. We need to consider making the PAC public. I'll discuss internally. For now, we should be able to merge the current PR. I checked locally that all these workflows pass.
